### ISSUE TITLE
Basic authentication fails

### ISSUE THREAD

**Issue by LeMoussel** [sjdirect/abot#122]
I want to crawl site that have a simple .htaccess file for basic authentication.
I set configuration parameters to :
```
isAlwaysLogin="true" 
loginUser="TheUserName"
loginPassword="ThePassword"
```
But I can't crawl site.  I get a 401 (not authenticated error).

In `PageRequester.cs`, there is
```
protected virtual HttpWebRequest BuildRequestObject(Uri uri)
{
....
            if (_config.IsAlwaysLogin)
            {
                request.Credentials = new NetworkCredential(_config.LoginUser, _config.LoginPassword);
                request.UseDefaultCredentials = false;
            }

            return request;
}
```
This works ONLY if the server returns 401 first. The client DOES NOT send credentials on first request
ONLY after a 401.

So use, I modify this to send credentials right away (old fashioned way)
```
protected virtual HttpWebRequest BuildRequestObject(Uri uri)
{
....
            if (_config.IsAlwaysLogin)
            {
                string credentials = Convert.ToBase64String(System.Text.Encoding.ASCII.GetBytes(_config.LoginUser + ":" + _config.LoginPassword));
                request.Headers[HttpRequestHeader.Authorization] = "Basic " + credentials;
            }

            return request;
}
```

**Comment by draperj** [comment]
Hello,

When running abot against a Windows Authenticated ASP website, the commit mentioned breaks proper authentication. 

When I reversed the code in the main branch to the previous authentication resolution method I was able to properly authenticate with the website. 

Would it be viable to have a .config switch to elect the possible authentication method?

**Comment by sjdirect** [comment]
@LeMoussel, what do you think of @draperj comments? This was a contribution so I have never used it for basic auth.

**Comment by LeMoussel** [comment]
Perhaps with the Windows Authenticated ASP website, the server returns 401 first.

**Comment by draperj** [comment]
That is correct, a 401 is generated when using the newer authenticated method along with the username configured in the .config.

**Comment by sjdirect** [comment]
Can one of you submit a pull request with a solution that solves both issues or is this a one or the other kind of situation?

**Comment by ramazancesur** [comment]
İ dont  know this submit

İ dont understand you can you get information for me

25 Mayıs 2016 Çarşamba tarihinde, Steven <notifications@github.com> yazdı:

> Can one of you submit a pull request with a solution that solves both
> issues or is this a one or the other kind of situation?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/sjdirect/abot/issues/122#issuecomment-221687438>
>


-- 
--
Saygılarımla

Ramazan CESUR

BİLGİSAYAR MÜHENDİSİ

ANKARA

**Comment by sjdirect** [comment]
Just added a minor change to this available in nuget version 1.6.0.20.
